src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/DefaultRazorFormattingService.cs

```diff
@@ -0,0 +1,287 @@
+using Microsoft.AspNetCore.Razor.Language.Legacy;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Text;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using OmniSharp.Extensions.LanguageServer.Protocol.Server;
+using Range = OmniSharp.Extensions.LanguageServer.Protocol.Models.Range;
+using RoslynFormattingOptions = Microsoft.CodeAnalysis.Formatting.FormattingOptions;
+using OmniSharpFormattingOptions = OmniSharp.Extensions.LanguageServer.Protocol.Models.FormattingOptions;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Formatting
+{
+    internal class DefaultRazorFormattingService : RazorFormattingService
+    {
+        private readonly ForegroundDispatcher _foregroundDispatcher;
+        private readonly RazorDocumentMappingService _documentMappingService;
+        private readonly FilePathNormalizer _filePathNormalizer;
+        private readonly ProjectSnapshotManagerAccessor _projectSnapshotManagerAccessor;
+        private readonly ILanguageServer _server;
+
+        public DefaultRazorFormattingService(
+            ForegroundDispatcher foregroundDispatcher,
+            RazorDocumentMappingService documentMappingService,
+            FilePathNormalizer filePathNormalizer,
+            ProjectSnapshotManagerAccessor projectSnapshotManagerAccessor,
+            ILanguageServer server)
+        {
+            if (foregroundDispatcher is null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            if (documentMappingService is null)
+            {
+                throw new ArgumentNullException(nameof(documentMappingService));
+            }
+
+            if (filePathNormalizer is null)
+            {
+                throw new ArgumentNullException(nameof(filePathNormalizer));
+            }
+
+            if (projectSnapshotManagerAccessor is null)
+            {
+                throw new ArgumentNullException(nameof(projectSnapshotManagerAccessor));
+            }
+
+            if (server is null)
+            {
+                throw new ArgumentNullException(nameof(server));
+            }
+
+            _foregroundDispatcher = foregroundDispatcher;
+            _documentMappingService = documentMappingService;
+            _filePathNormalizer = filePathNormalizer;
+            _projectSnapshotManagerAccessor = projectSnapshotManagerAccessor;
+            _server = server;
+        }
+
+        public override Task<TextEdit[]> FormatAsync(Uri uri, RazorCodeDocument codeDocument, Range range, OmniSharpFormattingOptions options)
+        {
+            var syntaxTree = codeDocument.GetSyntaxTree();
+            var formattingSpans = syntaxTree.GetFormattingSpans();
+            var indentations = GetLineIndentationMap(codeDocument.Source, formattingSpans);
```

Comment:
Yes, I agree. My plan was to do that once we have the need to merge results of multiple formatters into one. The general flow will be,
Original sourcetext -> Format razor blocks -> apply edits and reparse -> Format special HTML parts -> apply edits and reparse -> Format C# -> remap and apply edits and reparse -> Final sourcetext -> return diff of original and final sourcetext
So, having formatters that know format parts it knows about seem like a good fit.
I haven't done that yet because we don't have multiple formatters yet.

Reply:
Cooo coo, sounds good!